tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,466 @@
+    T mutable_device_data() {
+	fromDevice(device_);
+        return static_cast<T>(*access_data(0));
+    }*/
+
+    size_t size() const {
+        size_t new_size = 1;
+        for (auto d : shape_) { new_size *= d; }
+        return new_size;
+    }
+
+    /* @brief Fills all the tensor values with a given value
+     *
+     * @param value The value to fill the tensor
+     */
+    void fill(const U value) {
+        std::fill(host_data_->begin(), host_data_->end(), value);
+    }
+
+    /* @brief Element-wise addition
+     */
+    Tensor<U> add(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) + src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise addition
+     */
+    Tensor<U> add(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) + scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise subtraction
+     */
+    Tensor<U> sub(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) - src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise subtraction
+     */
+    Tensor<U> sub(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) - scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise multiplication
+     */
+    Tensor<U> mul(const Tensor<U>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) * src[i];
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise multiplication
+     */
+    Tensor<U> mul(const float_t scalar) const {
+        Tensor<U> res(this->shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) * scalar;
+        });
+
+        return std::move(res);
+    }
+
+    /* @brief Element-wise division
+     */
+    Tensor<U> div(const Tensor<>& src) const {
+        if (this->shape() != src.shape()) {
+            throw nn_error("Tensor must have same shape");
+        }
+
+        Tensor<U> res(src.shape());
+
+        for_i(true, res.size(), [&](size_t i) {
+            res[i] = this->operator[](i) / (src[i] + 1e-10);
```

Comment:
@edgarriba you fixed one problem out of two. To avoid division by zero you should explicitly check if a number isn't zero. I think you should add assert on equality to zero to not hurt performance